### PR TITLE
Reconcile check.failed against current HEAD, tame cancellation churn

### PR DESCRIPTION
**Pull request by jdtzmn**
## Problem

Every push starts fresh CI on the new commit and GitHub cancels in-flight runs on the previous one. premind's watcher stores a `check.failed` event for each failing/cancelled check-run it sees, but never reconciles it against the PR's current HEAD before surfacing it. An agent that has already pushed a fix keeps getting "Action required" reminders for checks that ran on commits it already replaced — each one costing a verify-and-dismiss turn (`git rev-parse HEAD` + `gh api .../check-runs`) for zero code outcome. Concurrency-cancelled runs and aggregate gate jobs failing due to a cancelled prerequisite compound this, and matrix/shard jobs emit noise per-shard as they complete.

## What changed

Five independently-reviewable commits:

1. **Tag check events with head SHA and recognize cancelled checks** — every `check.*` event payload now carries the `headSha` it was observed on, and a GitHub check-run conclusion of `CANCELLED` is now mapped to a new `check.cancelled` kind (low priority) instead of falling through to `check.created`.

2. **Reconcile check.failed against current HEAD, rewrite reminder text** — `buildReminderBatch` looks up the PR's current snapshot head SHA and partitions `check.failed`/`check.cancelled` events into live (matches HEAD, or no headSha recorded — backward compatible) vs superseded (stale). Superseded events are excluded from the actionable-blocker check and coalesced into one summary line per old commit instead of repeating individually. Reminder text is also tightened: the header states the qualified PR + HEAD SHA once, live changes and superseded history render as separate sections, and the action-required line is shorter.

3. **Suppress cancellation churn from races and aggregate-gate failures** — handles two sources of noise from concurrent runs on the *same* commit (a push and a label-triggered run racing): checks are grouped by name per tick so an active rerun always wins over a stale fail/cancelled duplicate, and a failing check is *conditionally* downgraded to `check.cancelled` when a sibling in the same workflow was cancelled this tick.

4. **Coalesce matrix/shard checks: quiet on green, loud on first red** — matrix jobs are recognized by stripping a trailing `(...)`/`[...]` segment from the check name. A failing shard always fires immediately and uncondensed. A shard succeeding is suppressed while siblings are still running, suppressed again if a sibling already failed, and rolled up into one "N/N shards succeeded" event only once the whole job finishes clean.

5. **Fix false-negative in aggregate-gate cancellation heuristic** — an independent review caught that commit 3's downgrade, as originally written, keyed only on workflow name and could mask a *real* failure: either a matrix shard failing for a genuine reason (fail-fast routinely cancels its siblings as a normal consequence) or an unrelated job failing for a genuine reason while something else in the same workflow was cancelled for an unrelated one. This commit narrows the heuristic to require the failing check's own name to look like an aggregate gate (`/\b(gate|required|fail[- ]?if|overall|rollup)\b/i`) **and** that it is not itself part of a matrix group — so a real matrix-shard failure is never eligible for the downgrade regardless of naming. Two regression tests reproduce the exact false-negative scenarios and assert they correctly stay `check.failed`/high priority.

## Testing

- `bun run --bun tsc -p tsconfig.json --noEmit` — clean
- `bun run test` — 264/264 passing (40 new tests added across the five commits)
- Reviewed by an independent Sonnet agent with no prior context on this branch, adversarially checking for logic bugs (esp. false negatives that could hide a real CI failure), backward compatibility, and test quality. Its one blocker finding is what commit 5 fixes; smaller nits (misleading comment, a documented edge case in cross-tick matrix-group stability, and a minor query-ordering inefficiency) are also addressed in commit 5.

_(Drafted by Jacob's coding agent on his behalf)_

### CODE REVIEW

File: src/daemon/github/diff.ts [hunk 1]
```diff
@@ -446,21 +449,100 @@ export function diffSnapshot(previous: PullRequestSnapshot | null, next: PullReq
...
+
+  // Matrix/shard jobs ("unit-tests (shard 1)", "test (ubuntu-latest)", ...) are
+  // detected by stripping a trailing "(...)"/"[...]" segment from the name;
+  // checks in the same workflow that reduce to the same base name are treated
+  // as one job's shards. Quiet on green: an individual shard turning green
+  // while siblings are still running is not worth waking the agent for, and
+  // is rolled up into one summary once the whole job finishes clean. Loud on
+  // first red: a failing shard is always reported immediately, uncondensed,
+  // so the agent can start fixing it without waiting on the rest.
+  const matrixBaseName = (name: string) => name.replace(/\s*[([][^()[\]]*[)\]]\s*$/, "").trim() || name
```
Comment (jdtzmn) [inline src/daemon/github/diff.ts:512]: What's the risk associated with false positives for these?

Comment (jdtzmn) [reply src/daemon/github/diff.ts:512]: Good catch, and yes — there was a real one here, which is exactly what an independent review agent I ran after opening this PR flagged as a blocker (before I saw this comment).

As originally written (the version you're commenting on, before commit `43053d5`), the downgrade fired for *any* `check.failed` if *any other check in the same workflow* was cancelled — no restriction to gate-style jobs. Concretely, two false positives:

1. **Fail-fast matrix cancellation**: GitHub Actions' default `fail-fast: true` cancels sibling shards when one shard genuinely fails. That real failure would get masked as `check.cancelled` purely because it shared a workflow with a cancelled sibling.
2. **Unrelated jobs in one workflow**: job A fails for a genuine reason, job B is cancelled for an unrelated reason, same workflow → job A's real failure gets masked the same way.

Both defeat the whole point of this PR (never hide a live failure), so I fixed it in `43053d5`: the downgrade now requires the failing check's own name to look like an aggregate gate (`/\b(gate|required|fail[- ]?if|overall|rollup)\b/i`) **and** that it isn't itself part of a matrix group at all — so a real matrix-shard failure is never eligible for the downgrade regardless of naming. Added two regression tests (`does not mask an unrelated job's real failure...` and `does not mask a real matrix-shard failure whose siblings were fail-fast cancelled...`) that reproduce both scenarios above and assert they correctly stay `check.failed`/high priority.

The remaining risk is narrower and intentional: a genuine standalone aggregate-gate job (name matches the pattern, not part of a matrix) whose failure happens to coincide with an unrelated cancellation elsewhere in the same workflow could still be misclassified. I judged that acceptable — it's a much smaller surface than "any job, any cancellation," and the alternative (correlating via an actual run/check-suite dependency graph) isn't available from the data GitHub's REST/GraphQL check-run API exposes here. Documented as a best-effort heuristic in the code comment.

_(Drafted by Jacob's coding agent on his behalf)_
